plugins/hu.bme.mit.massif.simulink.layout.elk/src/hu/bme/mit/massif/simulink/layout/elk/ElkLayoutProvider.java

```diff
@@ -0,0 +1,143 @@
+    @Override
+    public void updateSpecifications(Map<Block, BlockLayoutSpecification> layout) {
+        ElkNode graph = createGraph(layout.keySet());
+        layout(graph);
+        layout.putAll(getSpecifications(graph));
+    }
+
+    private static ElkNode createGraph(Set<Block> blocks) {
+        ElkNode graph = ElkGraphUtil.createGraph();
+        
+        graph.setProperty(CoreOptions.ALGORITHM, LayeredOptions.ALGORITHM_ID);
+        
+        graph.setProperty(CoreOptions.DIRECTION, Direction.RIGHT);
+        
+        Collection<ElkNode> nodes = blocks.stream().map(block -> createNode(block, graph)).collect(Collectors.toSet());
+
+        for (Block block : blocks) {
+            for (OutPort outPort : block.getOutports()) {
+                createEdge(outPort.getConnection(), nodes);
+            }
+        }
+        
+        return graph;
+    }
+
+    private final static IProperty<Block> BLOCK = new Property<Block>("hu.bme.mit.massif.simulink.layout.elk.block");
+
+    private static ElkNode createNode(Block block, ElkNode graph) {
+        ElkNode node = ElkGraphUtil.createNode(graph);
+        node.setProperty(BLOCK, block);
+        node.setWidth(10);
+        node.setHeight(10);
+        return node;
+    }
+    
+    private static ElkEdge createEdge(Connection connection, Iterable<ElkNode> nodes) {
+        Iterable<ElkNode> sources = findNodes(connection.getFrom().getContainer(), nodes);
+        Collection<ElkConnectableShape> sourcePorts = createPorts(sources);
+        
+        Iterable<Block> targetBlocks = Iterables.transform(getSingleConnections(connection), ElkLayoutProvider::getTarget);
+        Iterable<ElkNode> targets = Iterables.concat(Iterables.transform(targetBlocks, target -> findNodes(target, nodes)));
+        Collection<ElkConnectableShape> targetPorts = createPorts(targets);
+        
+        return ElkGraphUtil.createHyperedge(sourcePorts, targetPorts);
+    }
+
+    private static Collection<ElkConnectableShape> createPorts(Iterable<ElkNode> targets) {
+        return ImmutableSet.copyOf(Iterables.transform(targets, ElkGraphUtil::createPort));
+    }
+    
+    public static Iterable<SingleConnection> getSingleConnections(Connection connection) {
+        if (connection instanceof SingleConnection) {
+            SingleConnection singleConnection = (SingleConnection) connection;
+            return ImmutableSet.of(singleConnection);
+        } else if (connection instanceof MultiConnection) {
+            MultiConnection multiConnection = (MultiConnection) connection;
+            return multiConnection.getConnections();
+        } else {
+            return ImmutableSet.of();
+        }
+    }
+    
+    public static Block getTarget(SingleConnection singleConnection) {
+        return singleConnection.getTo().getContainer();
+    }
+
+    private static Iterable<ElkNode> findNodes(Block block, Iterable<ElkNode> nodes) {
+        return Iterables.filter(nodes, node -> getBlock(node) == block);
```

Comment:
I know we don't expect thousands of blocks, but wouldn't it be better to create a `Map<Block,ElkNode>` instead?